File: code/Python/Chapter3/3-2/singly_linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class SinglyLinkedList:
    def __init__(self):
        self.__head = None
        self.__tail = None
        self.__len = 0
    
    def is_empty(self):
        return self.__head == None
    
    def __len__(self):
        return self.__len
    
    def add(self, data):
        node = Node(data)
        if self.is_empty():
            self.__head = node
        else:
            self.__tail.next = node
        self.__tail = node
        self.__len += 1
    
    def insert(self, index, data):
        if index < 0 or index > self.__len:
            raise IndexError("Index out of range")

        node = Node(data)
        if index == 0:
            node.next = self.__head
            self.__head = node
        else:
            cur = self.__head
            for _ in range(index - 1):
                cur = cur.next
            node.next = cur.next
            cur.next = node
        self.__len += 1
    
    def remove(self, index):
        if index < 0 or index >= self.__len:
            raise IndexError("Index out of range")

        if index == 0:
            data = self.__head.data
            self.__head = self.__head.next
        else:
            cur = self.__head
            for _ in range(index - 1):
                cur = cur.next
            data = cur.next.data
            cur.next = cur.next.next
        self.__len -= 1
        return data
    
    def get(self, index):
        if index < 0 or index >= self.__len:
            raise IndexError("Index out of range")

        cur = self.__head
        for _ in range(index):
            cur = cur.next
        return cur.data
    
    def set(self, index, data):
        if index < 0 or index >= self.__len:
            raise IndexError("Index out of range")

        cur = self.__head
        for _ in range(index):
            cur = cur.next
        cur.data = data
    
    def __str__(self):
        cur = self.__head
        s = "["
        while cur != None:
            s += str(cur.data)
            if cur.next != None:
                s += ", "
            cur = cur.next
        s += "]"
        return s
    
    def __iter__(self):
        self.__cur = self.__head
        return self
    
    def __next__(self):
        if self.__cur == None:
            raise StopIteration
        data = self.__cur.data
        self.__cur = self.__cur.next
        return data
    
    def __reverse_recursive(self, node):
        if node == None or node.next == None:
            return node

        new_head = self.__reverse_recursive(node.next)
        node.next.next = node
        node.next = None
        return new_head

    def reverse_recursive(self):
        self.__head = self.__reverse_recursive(self.__head)

    def reverse(self):
        if self.__head == None or self.__head.next == None:
            return

        cur = self.__head
        pre = None
        while cur != None:
            next = cur.next
            cur.next = pre
            pre = cur
            cur = next
        self.__head = pre

File: code/Java/Chapter3/3-2/SinglyLinkedList.java
import java.util.Iterator;

public class SinglyLinkedList<T> {
    private class Node<T> {
        T data;
        Node<T> next;

        Node(T elem) {
            this.data = elem;
            this.next = null;
        }
    }

    private Node<T> head;
    private Node<T> tail;
    private int size;

    public SinglyLinkedList() {
        head = null;
        tail = null;
        size = 0;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public int size() {
        return size;
    }

    public void clear() {
        head = null;
        tail = null;
        size = 0;
    }

    public void add(T elem) {
        Node<T> newNode = new Node<T>(elem);
        if (head == null) {
            head = newNode;
            tail = newNode;
        } else {
            tail.next = newNode;
            tail = newNode;
        }
    }

    public void add(int index, T elem) throws IndexOutOfBoundsException {
        if (index < 0 || index > size) {
            throw new IndexOutOfBoundsException("Index out of bounds");
        }

        Node<T> newNode = new Node<T>(elem);
        if (index == 0) {
            newNode.next = head;
            head = newNode;
        } else {
            Node<T> prev = head;
            for (int i = 0; i < index - 1; i++) {
                prev = prev.next;
            }
            newNode.next = prev.next;
            prev.next = newNode;
        }
        size++;
    }

    public T remove(int index) throws IndexOutOfBoundsException {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("Index out of bounds");
        }

        T elem;
        if (index == 0) {
            elem = head.data;
            head = head.next;
        } else {
            Node<T> prev = head;
            for (int i = 0; i < index - 1; i++) {
                prev = prev.next;
            }
            elem = prev.next.data;
            prev.next = prev.next.next;
        }
        size--;
        return elem;
    }

    public T get(int index) throws IndexOutOfBoundsException {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("Index out of bounds");
        }

        Node<T> current = head;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }
        return current.data;
    }

    public void set(int index, T elem) throws IndexOutOfBoundsException {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("Index out of bounds");
        }

        Node<T> current = head;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }
        current.data = elem;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        Node<T> current = head;
        while (current != null) {
            sb.append(current.data);
            if (current.next != null) {
                sb.append(", ");
            }
            current = current.next;
        }
        sb.append("]");
        return sb.toString();
    }

    public Iterator<T> iterator() {
        return new Iterator<T>() {
            private Node<T> current = head;

            @Override
            public boolean hasNext() {
                return current != null;
            }

            @Override
            public T next() {
                T elem = current.data;
                current = current.next;
                return elem;
            }
        };
    }

    private Node<T> reverseRecursive(Node<T> current) {
        if (current == null || current.next == null) {
            return current;
        }
        Node<T> newHead = reverse(current.next);
        current.next.next = current;
        current.next = null;
        return newHead;
    }

    public void reverseRecursive() {
        head = reverseRecursive(head);
    }

    public void reverse() {
        Node<T> prev = null;
        Node<T> current = head;
        Node<T> next = null;
        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        head = prev;
    }
}
